crates/rattler/src/libsolv/repo.rs

```diff
@@ -0,0 +1,118 @@
+        unsafe { self.0 .0.cast().as_ref() }
+    }
+}
+
+impl<'pool> DerefMut for Repo<'pool> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { self.0 .0.cast().as_mut() }
+    }
+}
+
+/// Wrapper type so we do not use lifetime in the drop
+struct RepoOwnedPtr(NonNull<ffi::Repo>);
+
+/// Destroy c-side of things when repo is dropped
+impl Drop for RepoOwnedPtr {
+    // Safe because we have coupled Repo lifetime to Pool lifetime
+    fn drop(&mut self) {
+        unsafe { ffi::repo_free(self.0.as_mut(), 1) }
+    }
+}
+
+impl RepoRef {
+    /// Converts from a pointer to an `ffi::Repo` to a reference of Self.
+    pub(super) fn from_ptr<'pool>(repo: NonNull<ffi::Repo>) -> &'pool Self {
+        // Safe because a `RepoRef` is a transparent wrapper around `ffi::Repo`
+        unsafe { std::mem::transmute(repo.as_ref()) }
+    }
+
+    /// Returns a pointer to the wrapped `ffi::Repo`
+    fn as_ptr(&self) -> NonNull<ffi::Repo> {
+        // Safe because a `RepoRef` is a transparent wrapper around `ffi::Repo`
+        unsafe { NonNull::new_unchecked(self as *const Self as *mut Self).cast() }
+    }
+
+    /// Returns a reference to the wrapped `ffi::Repo`.
+    fn as_ref(&self) -> &ffi::Repo {
+        // Safe because a `RepoRef` is a transparent wrapper around `ffi::Repo`
+        unsafe { std::mem::transmute(self) }
+    }
+
+    /// Returns the pool that created this instance
+    pub fn pool(&self) -> &PoolRef {
+        // Safe because a `PoolRef` is a wrapper around `ffi::Pool`
+        unsafe { &*(self.as_ref().pool as *const PoolRef) }
+    }
+
+    /// Add conda json to the repo
+    pub fn add_conda_json<T: AsRef<str>>(&mut self, json_path: T) -> anyhow::Result<()> {
+        let c_json = c_string(json_path.as_ref());
+        let mode = c_string("r");
+        unsafe {
+            // Cast needed because types do not match in bindgen
+            // TODO: see if lib types could be used in bindgen
+            // Safe because we check nullptr
+            let file = libc::fopen(c_json.as_ptr(), mode.as_ptr()) as *mut ffi::FILE;
+            if file.is_null() {
+                return Err(anyhow::anyhow!(
+                    "fopen returned a nullptr. '{}' does this file exist?",
+                    json_path.as_ref()
+                ));
+            }
+            // This line could crash if the json is malformed
+            let ret = ffi::repo_add_conda(self.as_ptr().as_mut(), file, 0);
+            if ret != 0 {
+                return Err(anyhow::anyhow!(
+                    "internal libsolv error while adding repodata to libsolv"
+                ));
+            }
+
+            // Libsolv needs this function to be called so we can work with the repo later
+            // TODO: maybe wolf knows more about this function
+            ffi::repo_internalize(self.as_ptr().as_mut());
+        }
+        Ok(())
+    }
+
+    /// Add a solvable to the Repo
+    pub fn add_solvable(&mut self) -> SolvableId {
```

Comment:
What is the purpose of this function? There doesn't seem to be a way to associate any data with the solvable.

Reply:
Removed it